src/crimson/os/seastore/seastore_types.h

```diff
@@ -1019,11 +1026,158 @@ inline extent_len_le_t init_extent_len_le(extent_len_t len) {
+ *
+ * [upgrade:1][object_info:76][object_content:51]
+ *
+ * Each pair of square brackets represents a property stored within the laddr_t
+ * interger. Each property contains a name and the count of bits it used.
+ *
+ * When the object info for different objects does not conflict within an OSD,
+ * we could obtain a useful property:
+ * Each rados object and its head/clone could have a unique laddr prefix within
+ * an OSD.
+ *
+ * Base on this property, we could:
+ * 1. Group different snapshots of a rados object to the same laddr prefix,
+ *    which could speed up cloning process.
+ * 2. If the rados object data/meta of a head/clone is under the same prefix,
+ *    removing a head/clone will also be possible via range deletion.
+ * 3. Track frequently accessed objects using laddr_t without keeping the full
+ *    object name.
+ *
+ * # Object Info
+ * The defination of this property is:
+ * [shard:6][pool:12][reverse_hash:16][local_object_id:42]
+ *
+ * The shard, pool and reverse_hash come from the information of rados object.
+ * The reverse_hash consists of pg_id(at most 16 bits) and the hash of object
+ * name, we only use pg_id part in laddr_t.
+ * The local_object_id is a random number to identify a unique object within
+ * seastore. Two different rados objects will never share the same object id
+ * within a pool.
+ *
+ * For global metadata logical extents, such as RootBlock, CollectionNode and
+ * SeastoreNodeExtent, these fields should always be zero, and the rados object
+ * data should never use this prefix.
+ *
+ * This layout allows:
+ * - 2^12=4096 pools per cluster
+ * - 2^6-1=63 shards per pool
+ * - 2^16=65536 pgs per pool and OSD
+ * - 2^42=4T objects per pg
+ *
+ * # Object Content
+ * Global meta logical extents use these bits as block address directly, while
+ * the rados objects futher divide them into:
+ * [local_clone_id:23][is_metadata:1][blocks:27]
+ *
+ * Like local_object_id, each clone/snapshot of a rados object has an unique
+ * local_clone_id under the same object laddr prefix. When creating a new
+ * snapshot, taking a random local clone id as the new base address for snap
+ * object.
+ * The indirect mapping of clone objects could only store the local_clone_id
+ * of its intermediate key, see pladdr_t::build_laddr().
+ *
+ * The rest 28 bits are used to represent the address of concrete data extents.
+ * Each address represents one 4KiB block on disk.
+ *
+ * is_metadata is true indicates the remaining bits represent the address of
+ * Omap*Node. Take a random value as address when allocating a new omap extent.
+ *
+ * TODO: It's possiable for OMapInnerNode store only the value of this field
+ * instead of a full 128-bit laddr_t to increase the fan-out of the omap node.
+ *
+ * When is_metadata is false, the remaining bits represent the address of
+ * ObjectDataBlock.
+ *
+ * This layout allows:
+ * 1. 2^23/2=4M clones per object(taking a snapshot will consume 2 clone ids)
```

Comment:
Why taking a snapshot will consume 2 clone ids?

Reply:
Taking snapshots is a symmetric process for both the clone onode and the head onode. The two new base laddr of them will reference the old head address space.

Reply:
OK, I think the basic version needs 2 clone ids per clone operation.

In a longer term, probably it's possible to save clone ids from the unmodified clones.

Reply:
> it's possible to save clone ids from the unmodified clones.

What does "unmodified clones" refer to?

Reply:
Meaning it isn't modified since being cloned, I think such a clone doesn't need another dedicated clone id, right?

Reply:
> Meaning it isn't modified since being cloned,

Seems this case is handled by the SnapSet of the object in PG layer? When two clones is same, the seastore won't do clone process, and when cloning started in seastore, there must be changes between two clones in SnapSet.

Reply:
(extended discussion, probably sidetracked)

I meant this is seastore internal and is related to how local-clone-id are assigned. IIUC in a simple version, during a symmetric seastore clone operation, 2 new local-clone-ids will be generated immediately and assigned to the new clone and the head. Seems to me it might be possible to save the local-clone-ids by delaying the local-clone-id allocation until the clone/head is actually modified by the seastore user.